linux/hydrogen.1
.TH HYDROGEN "1" "October 02, 2026" "Hydrogen 1.2.7" "User Commands"
.SH NAME
Hydrogen \- an advanced drum machine and step sequencer.
.SH SYNOPSIS
.PP
.B hydrogen
[options] [file]
.SH DESCRIPTION
.PP
\fBHydrogen\fR is both a drum sequencer for pattern-based programming and a drum
synth. Its simple yet feature-rich design makes it a versatile companion, be it
for quick practice at home, professional usage as fully-fledged drum machine in
your studio, or drum looper on steroids for your live performance.
.PP
Features:
.br
- First-class \fBMIDI\fR support
.br
- Highly customizable
.br
- Stable and performant
.br
- Loads of drumkits to choose from
.br
- In-depth control of instrument, sample, and note properties
.br
- High quality time-stretching and pitch shifting
.br
- Automate it using \fBMIDI\fR or \fBOSC\fR API
.SH OPTIONS
.TP
\fB\-d\fR, \fB\-\-driver\fR=[\fIAUDIO_DRIVER\fR]
Use the selected audio driver. The description of this option in the -\-help output contains all audio
drivers available on your system.
.TP
\fB\-s\fR, \fB\-\-song\fR=\fIFILE\fR
Load a song (\fI*.h2song\fR) at startup. Alternatively, you can just pass the filename without this option.
.TP
\fB\-p\fR, \fB\-\-playlist\fR=\fIFILE\fR
Load a playlist (\fI*.h2playlist\fR) at startup.
.TP
\fB\-k\fR, \fB\-\-kit\fR=\fIDRUMKIT_NAME\fR
Load a drumkit at startup.
.TP
\fB\-i\fR, \fB\-\-install\fR=\fIFILE\fR
Install a drumkit (\fI*.h2drumkit\fR).
.TP
\fB\-O\fR, \fB\-\-osc\-port\fR=\fIPORT\fR
Use custom OSC port.
.TP
\fB\-P\fR, \fB\-\-data\fR=\fIPATH\fR
Use an alternate system data path.
.TP
\fB\-\-user\-data\fR=\fIPATH\fR
Use an alternate user data path.
.IP
The default one is \fB$HOME\fR/.hydrogen.
.TP
\fB\-\-config\fR=\fIPATH\fR
Use an alternate config path.
.IP
The default one is \fB$HOME\fR/.hydrogen/hydrogen.conf.
.TP
\fB\-\-layout\fR=\fILAYOUT\fR
UI layout. Either \fItabbed\fR or \fIsingle\fR (default).
.TP
\fB\-n\fR, \fB\-\-nosplash\fR
Hide splash screen.
.TP
\fB\-V\fR, \fB\-\-verbose\fR=[\fILEVEL\fR]
Print a lot of debugging info
.IP
Level, if present, may be \fINone\fR, \fIError\fR, \fIWarning\fR, \fIInfo\fR, or \fIDebug\fR.
.TP
\fB\-L\fR, \fB\-\-log\-file\fR=\fIFILE\fR
Use custom log file.
.TP
\fB\-T\fR, \fB\-\-log\-timestamps\fR
Add timestamps to all log messages.
.TP
\fB\-\-log\-colors\fR
Use ANSI colors in log messages (default).
.TP
\fB\-\-no\-log\-colors\fR
Suppress ANSI colors in log messages.
.TP
\fB\-v\fR, \fB\-\-version\fR
Show version info.
.TP
\fB\-h\fR, \fB\-\-help\fR
Show this help message.
.SH FILES
.TP
\fB$HOME\fR/.hydrogen/data/drumkits
User-level drumkit folder. All drumkits you download via the online import as well as those you created manually will end up here. A portable version of a drumkit - the \fI.h2drumkit\fR tar archive file - can included into this folder using the \-\-install option. Only those kits residing either in this folder or in the system drumkit folder (for those kits you received via your package manager) will show up in the Sound Library.
.TP
\fB$HOME\fR/.hydrogen/data/patterns
If you save a pattern, it will be exported as \fI.h2pattern\fR XML file into this folder. Only the patterns residing in here will show up in the Sound Library.
.TP
\fB$HOME\fR/.hydrogen/data/playlists
This folder may have been original designed for holding \fI.h2playlist\fR XML files. But it does not serve any special purpose anymore and you can put them where ever you want.
.TP
\fB$HOME\fR/.hydrogen/data/plugins
In here you can put \fBLADSPA\fR plugins to be used by Hydrogen, which are not within \fBLADSPA_PATH\fR.
.TP
\fB$HOME\fR/.hydrogen/data/scripts
Playlists in Hydrogen can be configured to run a \fBBASH\fR script before loading a particular song. This folder may have been original designed for holding those script but you can put them where ever you want.
.TP
\fB$HOME\fR/.hydrogen/data/songs
If you want a \fI.h2song\fR XML file to show up in the Sound Library, you have to store it in here.
.SH CONFIGURATION
.TP
\fB$HOME\fR/.hydrogen/hydrogen.conf
XML file containing all configurations for Hydrogen.
.TP
\fB$HOME\fR/.hydrogen/data/emptySong.h2song
Whenever you create a new song, it will be derived from this template.
.TP
\fB$HOME\fR/.hydrogen/data/themes/
You can change colors, fonts, and many more in Hydrogen. These custom settings can be stored as \fI.h2theme\fR XML files in here by exporting the current theme in the Preferences dialog.
.SH ENVIRONMENT
.PP
.B LADSPA_PATH
.HP 
Path to \fBLADSPA\fR plugins.
.SH BUGS & FEATURES REQUEST
.PP
If you find any bug or if you would like to propose a new feature, please report it to https://github.com/hydrogen-music/hydrogen/issues
.SH AUTHORS
.PP
The hydrogen development team. See complete list at Hydrogen's "About" Dialog. You can contact them at <hydrogen-devel@lists.sourceforge.net>.
.TP
Subscription to the mailing list can be done here :
.br
http://lists.sourceforge.net/lists/listinfo/hydrogen-devel .
.SH SEE ALSO
A CLI tool for exporting songs and batch processing drumkits
.RS
\fBh2cli\fR(1)
.RE
.PP
A player Hydrogen-based songs on the command line
.RS
\fBh2player\fR(1)
.RE
.SH COPYRIGHT AND LICENSE
\fBCopyright (C)\fR 2002\-2008 Alessandro Cominu
.br
\fBCopyright (C)\fR 2008\-2026 The hydrogen development team
.PP
\fBHydrogen\fR is free software; you can redistribute it and/or modify it under the terms of the GPL version 2 or later.


NEWS.md
# 1.0.0-beta1*8000
- Usage of `auto` (in those contexts the type can be inferred by the
  object name, which is consistently used through the code), more
  consistent naming, and explicit `nullptr` instead of `NULL` in
  /src/core/src/hydrogen.cpp and some other files.
- Removing unused (obsolete) `Synth::m_pAudioOutput` and
  `Synth::setAudioOutput()`. Only the corresponding object and
  function of the `Hydrogen` class are used in the code base.
- Renaming
  - class AudioOutput -> AudioDriver
  - class MidiOutput -> MidiDriverOutput
  - class MidiInput -> MidiDriverInput
  - /src/core/include/hydrogen/IO/AudioOutput.h -> 
	/src/core/include/hydrogen/IO/AudioDriver.h
  - /src/core/include/hydrogen/IO/MidiOutput.h -> 
    /src/core/include/hydrogen/IO/MidiDriverOutput.h
  - /src/core/include/hydrogen/IO/MidiInput.h -> 
    /src/core/include/hydrogen/IO/MidiDriverInput.h
  - /src/core/src/IO/midi_input.cpp ->
	/src/core/src/IO/midi_driver_input.cpp
  - /src/core/src/IO/midi_output.cpp ->
    /src/core/src/IO/midi_driver_output.cpp
  - Hydrogen::getAudioOutput() -> Hydrogen::getAudioDriver()
  - Hydrogen::setAudioOutput() -> Hydrogen::setAudioDriver()
  - Hydrogen::setMidiOutput() -> Hydrogen::setMidiDriverOutput()
  - Hydrogen::setMidiInput() -> Hydrogen::setMidiDriverInput()
  - Hydrogen::__song -> Hydrogen::m_pSong
  - m_pMidiDriver -> Hydrogen::m_pMidiDriverInput
  - m_pMidiDriverOut -> Hydrogen::m_pMidiDriverOutput
- Introducing `Hydrogen::setMidiDriverInput()`,
  `Hydrogen::getMidiDriverInput()`, `Hydrogen::setMidiDriverOutput()`,
  `Hydrogen::getMidiDriverOutput()`, `Hydrogen::getAudioDriver()`, and
  `Hydrogen::setAudioDriver()` functions.
- Moving the `m_pAudioDriver`, `m_pMidiDriver`, and `m_pMidiDriverOut`
  objects into the `Hydrogen` class. 


src/www/hydrogen.php
<?php
/*
 * Hydrogen
 * Copyright(c) 2002-2008 by Alex >Comix< Cominu [comix@users.sourceforge.net]
 * Copyright(c) 2008-2024 The hydrogen development team [hydrogen-devel@lists.sourceforge.net]
 *
 * http://www.hydrogen-music.org
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY, without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program. If not, see https://www.gnu.org/licenses
 *
 */


	/*
		Use this script if you want to share your hydrogen songs,patterns and
		drumkits with other people. It generates a xml-file which can be used
		with the hydrogen soundlibrary.

		Just place this script in one directory with all your drumkits, songs
		and patterns and make it available via a webserver.

		Songs and patterns could be parsed in place. Since drumkits are archives,
		we cannot parse the files at runtime. Therefore, just the name and the url of the drumkit will be used.
		If you want additional information ( such as author / info ) take a look
		at the metaInfo.inc file (the file has to be in the same directory as this script).
		
	*/


	/*
		metaInfo.inc holds meta information about the drumkits, namely:
		
		- url
		- author
		- info
		- name

		If metaInfo.inc is not available, the script takes the filename as namely	and guesses the url
	*/

	if ( is_file("metaInfo.inc") ) {
		include("metaInfo.inc");
	}


	function getTag( $xml , $tag ){
		preg_match( "/". $tag . "(.*)</" , $xml , $found );
		if( ISSET( $found[1] ) ) {
			return $found[1];
		}
	}


	//$url should be defined in metaInfo.inc
	if( ! ISSET( $url ) ){
		$url = $_SERVER [ 'SERVER_NAME' ] ;
		$dir = dirname ( $_SERVER['PHP_SELF'] ) ;
		$url = "http://$url$dir";
		$drumkitImageUrl = $url + '/' + $drumkitImageDir;
	}

	$author = "";
	$info = "";

	/* Start of xml-document */

	print "<?xml version='1.0' encoding='UTF-8'?>\n";
	print "<drumkit_list>\n";


	/* Start of song listing */
	

	$dir = "./";
	if ($dh = opendir( $dir )) {
        	while (( $file = readdir($dh) ) !== false) {
			$extension = array_pop( explode(".", $file) );
			if( $extension == "h2song" ) {
				print "\t<song>\n";
				$content = file( $dir.$file );
				$xml = join( " ",$content );
				print "\t\t<name>" . getTag( $xml , "<name>" ) . "</name>\n";
				print "\t\t<url>" . $url.$file ."</url>\n";
				print "\t\t<author>" . getTag( $xml , "<author>" ) . "</author>\n";
				print "\t\t<info>" . getTag( $xml , "<info>" ) . "</info>\n";
				print "\t</song>\n";
			}
        	}
	    	closedir( $dh );
    	}
	/* End of song listing */



	/* Start of pattern listing */
	$dir = "./";
	if ($dh = opendir( $dir )) {
        	while ( ( $file = readdir( $dh ) ) !== false ) {
			$extension = array_pop(explode( ".", $file ));
			if( $extension == "h2pattern" ) {
				$content = file( $dir.$file );
				$xml = join( " " , $content );
				print "\t<pattern>\n";
				print "\t\t<name>" . getTag( $xml , "<name>" ) . "</name>\n";
				print "\t\t<url>" . $url.$file ."</url>\n";
				print "\t</pattern>\n";
			}
        	}
	    	closedir( $dh );
    	}	
	/* End of pattern listing */


	/* Start of drumkit listing */
	$dir = "./";
	$hasImages = false;
	if ($imageDirHandle = opendir( $drumkitImageDir ) ) {
		$hasImages = true;
		$imageFiles[] = array();
		while ( ( $f = readdir($imageDirHandle ) ) !== false ) {
		      $imageFiles[] = $file;
		}
		closedir($dh);
	}

	if ($dh = opendir( $dir ) ) {
        	while ( ( $file = readdir( $dh ) ) !== false) {
			$extension = array_pop( explode( ".", $file ) );
			if( $extension == "h2drumkit" ) {
				$content = file( $dir.$file );
				$xml = join( " " , $content );
				print "\t<drumkit>\n";
			
				//name: filename without extension
				$name = basename( $file,".h2drumkit" );

				If( ISSET( $drumkit_list[ $name ]) ){
					$author = $drumkit_list[ $name ][ "author" ];
					$info = $drumkit_list[ $name ][ "info" ];
				}

				print "\t\t<name> $name </name>\n";
				print "\t\t<url>" . $url.$file ."</url>\n";
				print "\t\t<author>$author</author>\n";
				print "\t\t<info>$info</info>\n";
				// Get image
				// Possible method: Have image filename same as basename of drumkit file.
				// ie: for DeathMetal.h2drumkit, image file would be DeathMetal.jpeg
				//     and allow for .png as well (non-case sensitive extension)
				//	The drumkit filenames need to be unique, and this would force the
				//     images to be unique names but be consistent.
				//     use regex like: /.*\.(?:jpeg|jpg|png)/i
				//
				//	Go through all files in image directory named $basename.* and
				//	see if there's a regex match, if so use that image file
				if ( $hasImages )
				{
					// search $imageFiles array for basename that matches with extension that matches regex					
					$images = preg_grep( "/.*\.(?:jpeg|jpg|png)/i", $imageFiles );
					if ( count( $images ) ) {
						// if there's more than one match we'll just use the first
						print "\t\t<image>" . $drumkitImageUrl . "/" . $images[0] . "</image>";
					}
				}
				print "\t</drumkit>\n";
			}
        	}
	    	closedir( $dh );
    	}	
	/* End of pattern listing */


	print "</drumkit_list>\n";
		

?>
